src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/NuGetSdkResolver.cs

```diff
@@ -85,20 +85,7 @@ public override SdkResult Resolve(SdkReference sdkReference, SdkResolverContext
                 return false;
             }
 
-            Dictionary<string, string> msbuildSdkVersions;
-
-            // Get the SDK versions from a previous state, otherwise find and load global.json to get them
-            if (context.State is Dictionary<string, string> dictionary)
-            {
-                msbuildSdkVersions = dictionary;
-            }
-            else
-            {
-                msbuildSdkVersions = GlobalJsonReader.GetMSBuildSdkVersions(context);
-
-                // Save the SDK versions in case this resolver is called again for another SDK in the same build
-                context.State = msbuildSdkVersions;
-            }
+            Dictionary<string, string> msbuildSdkVersions = GlobalJsonReader.GetMSBuildSdkVersions(context, out bool _);
```

Comment:
This is the most meaningful part of the change.  Rather than rely on the state from MSBuild which is not thread-safe, we now parse versions from `global.json` and cache the result based on the file path and last write time.  This means the file is only parsed once per instance so Visual Studio won't read it over and over.